File: p011.c
#include <stdio.h>
#include <stdint.h>

#define GRID_WIDTH 20
#define GRID_SIZE GRID_WIDTH * GRID_WIDTH

uint_fast8_t numbers[GRID_SIZE];
uint_fast32_t maximum = 0;

void update_maximum(int i, int step) {
	uint_fast32_t product = 1;
	for(int j = i; j != (i + 4 * step); j += step)
		product *= numbers[j];
	if(product > maximum)
		maximum = product;
}

int main() {
	// Read input file
	FILE *input_file;
	input_file = fopen("p011.input", "r");
	uint_fast8_t *current = numbers;
	while(fscanf(input_file, "%hhu", current++) == 1);
	fclose(input_file);

	// Find maximum
	for(int i = 0; i < GRID_SIZE; i++) {
		if(i % GRID_WIDTH <= (GRID_WIDTH - 4)) {
			update_maximum(i, +1); // horizontal right
			if(i >= 3 * GRID_WIDTH) update_maximum(i, -GRID_WIDTH + 1); // diagonal right up
			if(i < (GRID_WIDTH - 3) * GRID_WIDTH) update_maximum(i, +GRID_WIDTH + 1); // diagonal right down
		}
		if(i < (GRID_WIDTH - 3) * GRID_WIDTH) update_maximum(i, +GRID_WIDTH); // vertical down
	}

	// Print out result
	printf("%lu\n", maximum);
}	

File: p011.rs
use std::fs::File;
use std::io::prelude::*;

fn read_file() -> [u8; 400] {
    let mut file = File::open("p011.input").expect("Failed to open file");
    let mut contents = String::new();
    file.read_to_string(&mut contents).expect("Failed to read file");
    let mut array = [0u8; 400];
    for (i, w) in contents.split_whitespace().enumerate() {
        array[i] = w.parse().expect("Unable to parse integer");
    }
    array
}

#[test]
fn test_read_file() {
    let numbers = read_file();
    assert_eq!(numbers.len(), 20 * 20);
    assert_eq!(numbers[0], 8); // First number
    assert_eq!(numbers[399], 48); // Last number
}

fn get_product(numbers: &[u8], start: usize, step_by: usize) -> u32 {
    let stop = start + 3 * step_by + 1;
    let index_iterator = (start..stop).step_by(step_by);
    let product = index_iterator.map(|i| numbers[i]).fold(1u32, |acc, x| acc * (x as u32));
    product
}

#[test]
fn test_get_product_horizontal() {
    let numbers = read_file();
    let result = get_product(&numbers, 0, 1);
    assert_eq!(result, 08 * 02 * 22 * 97);
}

#[test]
fn test_get_product_vertical() {
    let numbers = read_file();
    let result = get_product(&numbers, 0, 20);
    assert_eq!(result, 08 * 49 * 81 * 52);
}

fn get_maximum_product(numbers: &[u8]) -> u32 {
    const GRID_WIDTH: usize = 20;
    let mut result = 0u32;
    for i in 0..numbers.len() {
        let column = i % GRID_WIDTH;
        if column < (GRID_WIDTH - 3) {
            result = result.max(get_product(numbers, i, 1)); // horizontal right
        }
        if i < (GRID_WIDTH - 3) * GRID_WIDTH {
            result = result.max(get_product(numbers, i, GRID_WIDTH)); // vertical down
            if column >= 3 {
                result = result.max(get_product(numbers, i, GRID_WIDTH - 1)); // diagonal left down
            }
            if column < (GRID_WIDTH - 3) {
                result = result.max(get_product(numbers, i, GRID_WIDTH + 1)); // diagonal right down
            }
        }
    }
    result
}

fn main() {
    let numbers = read_file();
    let solution = get_maximum_product(&numbers);
    println!("{}", solution);
}